src/sumlog.jl

```diff
@@ -1,42 +1,94 @@
 """
-$(SIGNATURES)
+    sumlog(X::AbstractArray{T}; dims)
 
-Compute `sum(log.(X))` with a single `log` evaluation.
+Compute `sum(log.(X))` with a single `log` evaluation,
+provided `float(T) <: AbstractFloat`.
 
-This is faster than computing `sum(log.(X))` or even `sum(log, X)`, in
-particular as the size of `X` increases.
-
-This works by representing the `j`th element of `X` as ``x_j = a_j  2^{b_j}``,
+This is faster than computing `sum(log, X)`, especially for large `X`.
+It works by representing the `j`th element of `X` as ``x_j = a_j  2^{b_j}``,
 allowing us to write
 ```math
 \\sum_j \\log{x_j} = \\log(\\prod_j a_j) + \\log{2} \\sum_j b_j
 ```
-Since ``\\log{2}`` is constant, `sumlog` only requires a single `log`
-evaluation.
 """
-function sumlog(x)
-    T = float(eltype(x))
-    _sumlog(T, values(x))
+sumlog(x::AbstractArray{T}; dims=:) where T = _sumlog(float(T), dims, x)
+
+function _sumlog(::Type{T}, ::Colon, x) where {T<:AbstractFloat}
+    sig, ex = mapreduce(_sumlog_op, x; init=(one(T), 0)) do xj
+        xj < 0 && Base.Math.throw_complex_domainerror(:log, xj)
+        float_xj = float(xj)
+        significand(float_xj), _exponent(float_xj) 
+    end
+    return log(sig) + IrrationalConstants.logtwo * T(ex)
 end
 
-@inline function _sumlog(::Type{T}, x) where {T<:AbstractFloat}
-    sig, ex = mapreduce(_sumlog_op, x; init=(one(T), zero(exponent(one(T))))) do xj
+function _sumlog(::Type{T}, dims, x) where {T<:AbstractFloat}
+    sig_ex = mapreduce(_sumlog_op, x; dims=dims, init=(one(T), 0)) do xj
+        xj < 0 && Base.Math.throw_complex_domainerror(:log, xj)
```

Comment:
I think it's better to not rely on some internal error function in base, it seems simple enough to throw a custom error message.
```suggestion
        xj < zero(xj) && throw(DomainError(xj, "log requires a non-negative argument"))
```